Mathlib/CategoryTheory/FiberedCategory/BasedCategory.lean

```diff
@@ -0,0 +1,269 @@
+
+end
+
+end BasedFunctor
+
+
+/-- A `BasedNatTrans` between two `BasedFunctor`s is a natural transformation `α` between the
+underlying functors, such that for all `a : 𝒳`, `α.app a` lifts `𝟙 S` whenever `𝒳.p.obj a = S`. -/
+structure BasedNatTrans {𝒳 : BasedCategory.{v₂, u₂} 𝒮} {𝒴 : BasedCategory.{v₃, u₃} 𝒮}
+    (F G : 𝒳 ⥤ᵇ 𝒴) extends CategoryTheory.NatTrans F.toFunctor G.toFunctor where
+  isHomLift' : ∀ (a : 𝒳.obj), IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (toNatTrans.app a) := by aesop_cat
+
+namespace BasedNatTrans
+
+open BasedFunctor
+
+variable {𝒳 : BasedCategory.{v₂, u₂} 𝒮} {𝒴 : BasedCategory.{v₃, u₃} 𝒮}
+
+initialize_simps_projections BasedNatTrans (+toNatTrans, -app)
+
+section
+
+variable {F G : 𝒳 ⥤ᵇ 𝒴} (α : BasedNatTrans F G)
+
+@[ext]
+lemma ext (β : BasedNatTrans F G) (h : α.toNatTrans = β.toNatTrans) : α = β := by
+  cases α; subst h; rfl
+
+instance app_isHomLift (a : 𝒳.obj) : IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (α.toNatTrans.app a) :=
+  α.isHomLift' a
+
+lemma isHomLift {a : 𝒳.obj} {S : 𝒮} (ha : 𝒳.p.obj a = S) :
+    IsHomLift 𝒴.p (𝟙 S) (α.toNatTrans.app a) := by
+  subst ha; infer_instance
+
+end
+
+/-- The identity natural transformation is a `BasedNatTrans`. -/
+@[simps]
+def id (F : 𝒳 ⥤ᵇ 𝒴) : BasedNatTrans F F where
+  toNatTrans := CategoryTheory.NatTrans.id F.toFunctor
+  isHomLift' := fun a => of_fac 𝒴.p _ _ (w_obj F a) (w_obj F a) (by simp)
+
+/-- Composition of `BasedNatTrans`, given by composition of the underlying natural
+transformations. -/
+@[simps]
+def comp {F G H : 𝒳 ⥤ᵇ 𝒴} (α : BasedNatTrans F G) (β : BasedNatTrans G H) :
+    BasedNatTrans F H where
+  toNatTrans := CategoryTheory.NatTrans.vcomp α.toNatTrans β.toNatTrans
+  isHomLift' := by
+    intro a
+    rw [CategoryTheory.NatTrans.vcomp_app]
+    infer_instance
+
+@[simps]
+instance homCategory (𝒳 : BasedCategory.{v₂, u₂} 𝒮) (𝒴 : BasedCategory.{v₃, u₃} 𝒮) :
+    Category (𝒳 ⥤ᵇ 𝒴) where
+  Hom := BasedNatTrans
+  id := BasedNatTrans.id
+  comp := BasedNatTrans.comp
+
+@[ext]
+lemma homCategory.ext {F G : 𝒳 ⥤ᵇ 𝒴} (α β : F ⟶ G) (h : α.toNatTrans = β.toNatTrans) : α = β :=
+  BasedNatTrans.ext α β h
+
+end BasedNatTrans
+
+namespace BasedCategory
+
+open BasedFunctor BasedNatTrans
+
+section
+
+variable {𝒳 : BasedCategory.{v₂, u₂} 𝒮} {𝒴 : BasedCategory.{v₃, u₃} 𝒮}
+
+/-- The inverse of a based natural transformation whose underlying natural tranformation is an
+isomorphism. -/
+def mkNatIso {F G : 𝒳 ⥤ᵇ 𝒴} (α : F.toFunctor ≅ G.toFunctor)
+    (isHomLift' : ∀ a : 𝒳.obj, IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (α.hom.app a)) : F ≅ G where
+  hom := { toNatTrans := α.hom }
+  inv := {
+    toNatTrans := α.inv
+    isHomLift' := by
+      intro a
+      specialize isHomLift' a
+      rw [← NatIso.app_inv]
+      rw [← NatIso.app_hom] at isHomLift'
+      apply IsHomLift.lift_id_inv }
+
+/-- The inverse of a based natural transformation whose underlying natural tranformation carries an
+`IsIso` instance. -/
+noncomputable def mkNatIsIso {F G : 𝒳 ⥤ᵇ 𝒴} (α : F.toFunctor ⟶ G.toFunctor)
```

Comment:
Instead, could you prove that a based nat. trans is iso if the underlying nat. trans is. Then, `mkNatIsIso` would become basically unnecessary as it would essentially become a special of `asIso`.

Reply:
That's what I wanted to do originally, but I couldnt seem to get it to work. Both `[IsIso α.toNatTrans]` and `[IsIso (α.toNatTrans : F.toFunctor ⟶ G.toFunctor)]` give type errors (because `α.toNatTrans` is not recognized as a morphism in some category for some reason)

Reply:
Okay I figured out a trick :)

Reply:
I think it you be better to make it a lemma rather than an instance. (One approach would be to define a forget functor from based functors to functors, and phrase this by saying this functor reflects isomorphisms.)

Reply:
Why should it not be an instance? Will type-class inference not work? I will try to implement the approach with the forgetful functor now.

Reply:
I've left it there for now, but I'm happy to remove it if you think it's not useful.

Reply:
The "natural" direction is that if a based nat. trans is an iso, then the induced nat. trans is an iso: in this direction, this could certainly an instance. But, having instances in the other direction is problematic because of loops.

Reply:
Ohhhh I see! Thanks a lot! Then I will add an instance for the other direction instead